Blog: OpenGitOps 1.0 (#45)
open-gitops/website: content/blog/1.0-announcement/index.mdx

---
title: OpenGitOps 1.0 is finally here and why you should care

> date: "2021-11-03T09:00:01.284Z"

Author: ```suggestion
date: "2021-11-08T09:00:01.284Z"
```

description: "OpenGitOps 1.0 brings a mature standard to the DevOps industry."
author: Dan Garfield and Scott Rigby
thumbnail: images/GitOps-1.0.jpg
---

GitOps is a term coined by Alexis Richardson a few years ago and has grown immensely in popularity.
In a nutshell, GitOps is a set of principles for operating and managing software systems, derived from modern software operations, but also rooted in pre-existing and widely adopted best practices.

> However they differ in specific ways from traditional CI/CD pipeline approach that has dominated the industry for decades.

Reviewer: ```suggestion
However they differ in specific ways from traditional CI & CD pipeline approach that has dominated the industry for decades.
```
@todaywasawesome @scottrigby @murillodigital I agree with @countspongebob to not conflate CI and CD hence my suggestion here to change "CI/CD" to "CI & CD".

From https://www.youtube.com/watch?v=o2_sGFkYNvA&t=1184s by @countspongebob
> It is really critical to not conflate CI and CD.  I try hard to not use the phrase CI/CD because they are very different practices.  One of the lovely things about the GitOps model is the clean and understandable relationship that is created between CI as a Dev practice, and CD as an Ops practice.  I think the notion of the Immutability Firewall in between them based on Git is really excellent.

> GitOps is a term coined by Alexis Richardson a few years ago and has grown immensely in popularity. In a nutshell, GitOps offers a new pattern for deploying software that differs from the traditional CD pipeline approach that has dominated the industry for decades. Last year, Amazon, Azure, Codefresh, Github, Redhat, and Weaveworks teamed up to launch the GitOps Working Group under the CNCF. The goal was to formalize all the ideas that have developed around GitOps into a single cohesive set of principles that would be easily digestible and repeatable by any organization.

Reviewer: ```suggestion
GitOps is a term coined by Alexis Richardson a few years ago and has grown immensely in popularity. In a nutshell, GitOps offers a new pattern for deploying software that differs from the traditional CD pipelines approach that has dominated the industry for decades. Last year, Amazon, Azure, Codefresh, Github, Redhat, and Weaveworks teamed up to launch the GitOps Working Group under the CNCF. The goal was to formalize all the ideas that have developed around GitOps into a single cohesive set of principles that would be easily digestible and repeatable by any organization.
```
pipeline → pipelines to align with later usage of pipelines (plural)

Reviewer 2: Here CD pipeline is in agreement with "approach" which I think works better as a singular. Later the plural makes more sense.

  

The desired state of a GitOps managed system must be:

  

1. **Declarative**
A [system](https://github.com/open-gitops/documents/blob/main/GLOSSARY.md#software-system) managed by GitOps must have its desired state expressed [declaratively](https://github.com/open-gitops/documents/blob/main/GLOSSARY.md#declarative-description).

2. **Versioned and Immutable**
Desired state is [stored](https://github.com/open-gitops/documents/blob/main/GLOSSARY.md#state-store) in a way that enforces immutability, versioning and retains a complete version history.

3. **Pulled Automatically**
Software agents automatically pull the desired state declarations from the source.

4. **Continuously Reconciled**
Software agents [continuously](https://github.com/open-gitops/documents/blob/main/GLOSSARY.md#continuous) observe actual system state and [attempt to apply](https://github.com/open-gitops/documents/blob/main/GLOSSARY.md#reconciliation) the desired state.

  

## GitOps is more than just storing infrastructure-as-code

> As the [git history proves](https://github.com/open-gitops/documents/commits/main), the wording of each of these principles was very carefully chosen. While many of these principles should be familiar to DevOps professionals, many are surprised to learn that they still have work to do to implement GitOps fully.

Reviewer: @todaywasawesome I thought I had commented on this earlier, but maybe it got lost.

The terminology "DevOps professionals" feels awkward in this context about GitOps principles.

Some professionals who are led by DevOps principles of shared ownership, workflow automation, and rapid feedback, may even say that DevOps is a philosophy, and DevOps isn't a job title at all.

Since your target audience includes those professionals, I suggest not using the terminology "DevOps professionals" here.

Usage is similar to how "CI/CD" would be clearer when written as "CI & CD", to clarify that CI is a development-focused practice, and CD is an operations-focused practice.  CI & CD are separate practices.

If a blog says "DevOps professional" and "CI/CD", then the professionals who you are trying to reach may tune out the blog.

Semantically, a blog is clearer to the intended audience by writing "professional led by DevOps principles", and "CI & CD".

In stark contrast, if a blog's target audience are recruiters using an Applicant Tracking System (ATS) to search keywords, then an ATS system might index "DevOps professional" and "CI/CD" as score-able keywords.

```suggestion
As the [git history proves](https://github.com/open-gitops/documents/commits/main), the wording of each of these principles was very carefully chosen. While many of these principles should be familiar to professionals, many are surprised to learn that they still have work to do to implement GitOps fully.
```

Reviewer 2: Fair.

As [git history shows](https://github.com/open-gitops/documents/commits/v1.0.0), the wording of each principle and linked glossary item was very carefully chosen.

> While many of these principles should be familiar to DevOps professionals, many are surprised to learn that they still have work to do to implement GitOps fully.

Reviewer: ```suggestion
While many of these principles should be familiar to professionals, many are surprised to learn that they still have work to do to implement GitOps fully.
```
@scottrigby Would you accept this suggestion?  This suggestion from https://github.com/open-gitops/website/pull/45#discussion_r742245403 got lost in your rebase.

> Declarative is usually the easiest to understand. Not just infrastructure-as-code but all configuration data that is needed to run a system including the application layer. While this doesn’t generally include persistent application data such as database contents it generally includes credentials and configuration for data recovery or data access.

Reviewer: ```suggestion
Declarative is usually the easiest to understand. Not just infrastructure-as-code but all configuration data that is needed to run a system including the application layer. While this doesn’t generally include persistent application data such as database contents, it generally includes credentials and configuration for data recovery or data access.
```
comma; grammar

Not just infrastructure-as-code but all configuration data that is needed to run a system including the application layer.
While this doesn’t generally include persistent application data such as database contents it often includes credentials and configuration for data recovery or data access.

_Versioned and Immutable_ is often understood as “use git” but there is more to it than that.
If that desired state describes using a “latest” tag for example, it is no longer versioned because there is no way to rollback changes in the desired state.
Likewise, many version control systems can be used in GitOps as long as they meet those two basic requirements and teams use them in a conformant manner.

_Pulled Automatically_ means we have to have software agents constantly observing the desired state.
While triggers may exist to speed up that observation, a GitOps system shouldn’t exclusively rely on them.
Software agents need to constantly be aware of what the desired state should be, not only when a deliberate change is made.

> The verb “pull” here is used very clearly in contrast to the way traditional CI/CD pipelines function based on triggers that simply push.

Reviewer: ```suggestion
The verb “pull” here is used very clearly in contrast to the way traditional CI & CD pipelines function based on triggers that simply push.
```
@todaywasawesome @scottrigby @murillodigital I agree with @countspongebob to not conflate CI and CD hence my suggestion here to change "CI/CD" to "CI & CD".

From https://www.youtube.com/watch?v=o2_sGFkYNvA&t=1184s by @countspongebob
> It is really critical to not conflate CI and CD.  I try hard to not use the phrase CI/CD because they are very different practices.  One of the lovely things about the GitOps model is the clean and understandable relationship that is created between CI as a Dev practice, and CD as an Ops practice.  I think the notion of the Immutability Firewall in between them based on Git is really excellent.

> Finally, Continuously Reconciling brings it all together. The GitOps software agents have to be aware of the actual state of a system under management and attempt to apply the desired state. Being constantly aware of both the actual state and desired state means we can detect anytime they are out of alignment whether by changes to the desired state (normal operation) or changes to the actual state (drift), a GitOps software agent should detect this attempt to apply the desired state.

Reviewer: ```suggestion
Finally, Continuously Reconciling brings it all together. The GitOps software agents have to be aware of the actual state of a system under management and attempt to apply the desired state. Being constantly aware of both the actual state and desired state means we can detect anytime they are out of alignment whether by changes to the desired state (normal operation) or changes to the actual state (drift); a GitOps software agent should detect this attempt to apply the desired state.
```
semicolon; grammar

> When teams implement these principles and achieve GitOps they deploy more often, have fewer regressions, and are more competitive. They bring together decades of experience delivering software to create a standard that is incredibly powerful and accessible and rapidly being adopted by a growing number of open source tools and vendors.

Reviewer: ```suggestion
When teams implement these principles and achieve GitOps, they deploy more often, have fewer regressions, and are more competitive. They bring together decades of experience delivering software to create a standard that is incredibly powerful, accessible, and rapidly being adopted by a growing number of open source tools and vendors.
```
commas; grammar

## What’s next

With the publication of this standard we’re announcing two new initiatives to help more and more teams adopt these principles. First, “GitOps Blueprints” will be a place where anyone from the community can bring their stacks and share how they’ve implemented GitOps using various tools. Want to know how others have implemented GitOps using your favorite cloud, SCM, and orchestrator? GitOps Blueprints is a place where you’ll be able to see examples from people implementing GitOps in practice.

  

> Second, “GitOps Whitepapers” is a place to explore ideas around implementing the GitOps principles. For example, what if your SCM system becomes unavailable? When is it ok to “break glass” and start making manual changes? GitOps Whitepapers are meant to be a space where the community can share these ideas and document them.

Reviewer: ```suggestion
Second, “GitOps Green papers” is a place to explore ideas around implementing the GitOps principles. For example, what if your SCM system becomes unavailable? When is it ok to “break glass” and start making manual changes? GitOps Green papers are meant to be a space where the community can share these ideas and document them.
```

The description sounds more like Green papers than White papers.

According to https://en.wikipedia.org/wiki/White_paper
> By contrast, green papers, which are issued much more frequently, are more open-ended. Also known as consultation documents, green papers may merely propose a strategy to implement in the details of other legislation, or they may set out proposals on which the government wishes to obtain public views and opinion.

If App Delivery TAG GWG is a vendor in business-to-business marketing, then Whitepaper would make more sense.  However, App Delivery TAG GWG behaves more like a regulatory government body in the context of conformance.  Hence the suggestion to use Green papers.

Reviewer 2: Oh interesting, I've never heard of green papers before!

> White papers are a way the government can present policy preferences before it introduces legislation. Publishing a white paper tests public opinion on controversial policy issues and helps the government gauge its probable impact.

This sounds pretty close to what we're going for with these. Places people can bring policy preferences. I don't know though, what do you think @scottrigby @murillodigital ?

Reviewer 3: @todaywasawesome @scottrigby @murillodigital One reason I suggest green papers is because @todaywasawesome wrote `"to explore ideas"` which aligns with how green papers are more open-ended for public consultation.

For example:

• [**Know your green paper on marriages in South Africa**](http://www.dha.gov.za/index.php/notices/1449-know-your-green-paper-on-marriages-in-south-africa) at http://www.dha.gov.za/index.php/notices/1449-know-your-green-paper-on-marriages-in-south-africa

* [**May 2021: Green paper on marriages in South Africa**](https://www.gov.za/sites/default/files/gcis_document/202105/44529gon398.pdf) at https://www.gov.za/sites/default/files/gcis_document/202105/44529gon398.pdf

• [**Green Papers are documents published by the European Commission to stimulate discussion on given topics at European level**](https://eur-lex.europa.eu/summary/glossary/green_paper.html) at https://eur-lex.europa.eu/summary/glossary/green_paper.html

* [**January 2021: Green Paper on Ageing: Fostering solidarity and responsibility between generations**](https://eur-lex.europa.eu/legal-content/EN/TXT/?uri=CELEX%3A52021DC0050) at https://eur-lex.europa.eu/legal-content/EN/TXT/?uri=CELEX%3A52021DC0050

WDYT?  Thank you!

Reviewer: @todaywasawesome @scottrigby @murillodigital One reason I suggest green papers is because @todaywasawesome wrote `"to explore ideas"` which aligns with how green papers are more open-ended for public consultation.

For example:

• [**Know your green paper on marriages in South Africa**](http://www.dha.gov.za/index.php/notices/1449-know-your-green-paper-on-marriages-in-south-africa) at http://www.dha.gov.za/index.php/notices/1449-know-your-green-paper-on-marriages-in-south-africa

* [**May 2021: Green paper on marriages in South Africa**](https://www.gov.za/sites/default/files/gcis_document/202105/44529gon398.pdf) at https://www.gov.za/sites/default/files/gcis_document/202105/44529gon398.pdf

• [**Green Papers are documents published by the European Commission to stimulate discussion on given topics at European level**](https://eur-lex.europa.eu/summary/glossary/green_paper.html) at https://eur-lex.europa.eu/summary/glossary/green_paper.html

* [**January 2021: Green Paper on Ageing: Fostering solidarity and responsibility between generations**](https://eur-lex.europa.eu/legal-content/EN/TXT/?uri=CELEX%3A52021DC0050) at https://eur-lex.europa.eu/legal-content/EN/TXT/?uri=CELEX%3A52021DC0050

WDYT?  Thank you!

Author: I personally think this is super interesting. Let's just discuss as a larger group. Can move this idea to a GitHub discussion, and also add to the next meeting agenda, with a goal of getting more feedback into that discussion. that way this blog won't be held up on those details. Does that sound good?

> There’s so much more on the horizon and we need your help! [Get involved](https://opengitops.dev/get-involved)! We’d love more participation to build GitOps Blueprints, GitOps Whitepapers, plan GitOps events, and help with the direction of the Open GitOps projects.

Reviewer: ```suggestion
There’s so much more on the horizon and we need your help! [Get involved](https://opengitops.dev/get-involved)! We’d love more participation to build GitOps Blueprints, GitOps Green Papers, plan GitOps events, and help with the direction of the Open GitOps project.
```
See earlier re: green papers.

Coincidentally, Blue-Green deployments is a terminology in Software Engineering.  If we need a colorful mnemonic, this "GitOps Blueprints", "GitOps Green Papers", and Blue-Green deployments can be combined toward meaningful outcomes.

> Lets go!

Reviewer: ```suggestion
Let's go!
```
"Let's go" means let us go, whereas "Lets go" means allow us to go.

---
title: Announcing GitOps 1.0
date: "2021-10-11T09:00:01.284Z"
description: "Open GitOps 1.0 brings a mature standard to the DevOps industry."
author: Dan Garfield and ?
thumbnail: images/
---

> The GitOps Working Group has been working diligiently since announcing the formation of the group...

Reviewer: ```suggestion
The GitOps Working Group has been working diligently since announcing the formation of the group...
```

Fix minor typo: diligiently → diligently
